File: parser_and_lexer/lexer.cpp
#include "lexer.h"
#include <sstream>
#include <iostream>

using std::cin;
using std::cout;
using std::stringstream;

// Constructor
Lexer::Lexer()
{
    // Inserts reserved words into the id's table
    id_table["true"] = Id{Tag::TRUE, "true"};
    id_table["false"] = Id{Tag::FALSE, "false"};
    
    // Starts the input reading
    peek = cin.get();
}

int Lexer::Line_num()
{
    return line;
};

// Returns tokens to the parser
Token * Lexer::Scan()
{
    // Skips spaces, tabs and new lines
    while (isspace(peek))
    {
        if (peek == '\n')
	{
	    line += 1;
	};
	peek = cin.get();
    };

    // Returns the numbers
    if (isdigit(peek))
    {
        int v = 0;


	do
	{
	    // Converts char to int
	    int n = peek - '0';
	    v = 10 * v + n;
	    peek = cin.get();
        }
	while (isdigit(peek));

	// DEBUG: shows the recognized token
	cout << "<num, " << v << "> ";

	// Returns the NUM token
        token.n = Num{v};
	return &token.n;
    };

    // Returns the KW and id's
    if (isalpha(peek))
    {
        stringstream ss;

	do
	{
	    ss << peek;
	    peek = cin.get();
	}
	while (isalpha(peek));

	string s = ss.str();
	auto pos = id_table.find(s);

	// If the lexeme is already in the table
	if (pos != id_table.end())
	{
	    // Returns the ID token
	    token.i = pos->second;
            return &token.i;
	};

        // If the lexeme is not in the table
        Id new_id {Tag::ID, s};
	id_table[s] = new_id;

	// Returns the id token
        token.i = new_id;
	return &token.i;
    };

    // Operators and isolated alphanumeric characters
    Token op {peek};
    peek = ' ';

    // Returns the token
    token.t = op;
    return &token.t;
};

File: python_translator/lexer.py
class Tag:
    def __init__(self):
        self.ID = "ID"
        self.NUM = "NUM"
        self.TRUE = "TRUE"
        self.FALSE = "FALSE"
        self.EOF = "EOF"
        
class Token:
    def __init__(self, tag):
        self.tag = tag
        
    def str(self):
        return self.tag
        
class Id(Token):
    def __init__(self, tag, name):
        self.tag = tag
        self.name = name
        super().__init__(self.tag)
        
    def str(self):
        return self.name
        
class Num(Token):
    def __init__(self, value):
        self.value = value
        self.tag = Tag().NUM
        super().__init__(self.tag)
        
    def str(self):
        return str(self.value)
        
class TokenHandler:
    def __init__(self):
        self.t = None
        self.n = None
        self.i = None
        
class Lexer:
    def __init__(self, file_name):
        self.file = open(file_name, "r")
        self.line = 1
        self.peek = self.file.read(1)
        
        self.id_table = {}
        self.id_table["false"] = Id(Tag().FALSE, "false")
        self.id_table["true"] = Id(Tag().TRUE, "true")
        self.token = TokenHandler()
        
    def next_char(self):
        self.peek = self.file.read(1)
        
    def is_space(self):
        if self.peek == "\n":
            self.line += 1
        self.next_char()
        
    def is_digit(self):
        v = 0
        while self.peek.isdigit():
            v = 10 * v + int(self.peek)
            self.next_char()
            
        self.token.n = Num(v)
        return self.token.n
        
    def is_alpha(self):
        s = ''
        while self.peek.isalpha():
            s += self.peek
            self.next_char()
            
        value = self.id_table.get(s)
        if value is not None:
            self.token.i = value
            return self.token.i
        new_id = Id(Tag().ID, s)
        self.id_table[s] = new_id
        
        self.token.i = new_id
        return self.token.i
        
    def skip_comments(self):
        self.next_char()
        if self.peek == "/":
            while self.peek != "\n" and self.peek != "":
                self.next_char()
            self.next_char()
            return self.scan()
        else:
            self.token.t = Token("/")
            return self.token.t
        
    def lineno(self):
        return self.line
        
    def scan(self):
        while self.peek.isspace():
            self.is_space()
            
        if self.peek.isdigit():
            return self.is_digit()
            
        elif self.peek.isalpha():
            return self.is_alpha()
            
        elif self.peek == "/":
            return self.skip_comments()
                
        elif self.peek == '':
            return Token(Tag().EOF)
        
        else:
            symbol = Token(self.peek)
            peek = ' '
            self.next_char()
            self.token.t = symbol
            return self.token.t
